sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/LogicalQueryStageStrategy.scala

```diff
@@ -50,9 +52,34 @@ object LogicalQueryStageStrategy extends Strategy with PredicateHelper {
 
     case j @ Join(left, right, joinType, condition, _)
         if isBroadcastStage(left) || isBroadcastStage(right) =>
-      val buildSide = if (isBroadcastStage(left)) BuildLeft else BuildRight
-      BroadcastNestedLoopJoinExec(
-        planLater(left), planLater(right), buildSide, joinType, condition) :: Nil
+      def createBroadcastNLJoin() = {
+        val buildSide = if (isBroadcastStage(left)) BuildLeft else BuildRight
+        BroadcastNestedLoopJoinExec(
+          planLater(left), planLater(right), buildSide, joinType, condition) :: Nil
+      }
+
+      /**
+       * See. [SPARK-32290]
+       * Not in Subquery will almost certainly be planned as a Broadcast Nested Loop join,
+       * which is very time consuming because it's an O(M*N) calculation.
+       * But if it's a single column NotInSubquery, and buildSide data is small enough,
+       * O(M*N) calculation could be optimized into O(M) using hash lookup instead of loop lookup.
+       */
+      if (SQLConf.get.notInSubqueryHashJoinEnabled &&
+        joinType == LeftAnti &&
+        isBroadcastStage(right) &&
```

Comment:
Just a question: Is this check (`isBroadcastStage`) similar to `canBroadcastBySize` one above ? ie will `isBroadcastStage` apply the right size checks ?

Reply:
> Just a question: Is this check (`isBroadcastStage`) similar to `canBroadcastBySize` one above ? ie will `isBroadcastStage` apply the right size checks ?

I think two function both refer AUTO_BROADCASTJOIN_THRESHOLD correctly.
isBroadcastStage -> BroadcastQueryStageExec -> BroadcastExchangeExec -> AUTO_BROADCASTJOIN_THRESHOLD

canBroadcastBySize directly using AUTO_BROADCASTJOIN_THRESHOLD

Reply:
using canBroadcastBySize in unified pattern match